botocore/regions.py

```diff
@@ -10,89 +10,230 @@
+
+        :type allow_non_regional: bool
+        :param allow_non_regional: Set to True to include endpoints that are
+             not regional endpoints (e.g., s3-external-1,
+             fips-us-gov-west-1, etc).
+        :return: Returns an iterable of endpoint names (e.g., us-east-1).
+        """
+        raise NotImplementedError
+
 
-class EndpointResolver(object):
-    _CONSTRAINT_FUNCS = {
-        'startsWith': lambda x, y: str(x).startswith(y),
-        'notStartsWith': lambda x, y: not str(x).startswith(y),
-        'equals': lambda x, y: x == y,
-        'notEquals': lambda x, y: x != y,
-        'oneOf': lambda x, y: x in y
-    }
-    DEFAULT_SCHEME = 'https'
+class S3CompatResolver(BaseEndpointResolver):
+    """Adds S3 specific customizations to an endpoint resolver.
 
-    def __init__(self, rules):
-        self._rules = rules
+    This resolver proxies to another resolver and sets a default region
+    of us-east-1 for "s3".
+    """
+    def __init__(self, endpoint_resolver):
+        self._proxy = endpoint_resolver
 
-    def get_rules_for_service(self, service_name):
-        """Return the rules for a given service.
+    def list_endpoint_names(self, service_name, partition_name='aws',
+                            allow_non_regional=False):
+        return self._proxy.list_endpoint_names(
+            service_name, partition_name, allow_non_regional)
 
-        Note that the list returned is a reference to the actual list of
-        rules used for the service.  This list can be mutated to insert your
-        own heuristic rules.  You can also use a service name of ``_default``
-        to get the list of default rules that are applied if no service
-        name matches.
+    def construct_endpoint(self, service_name, region_name):
+        # Use us-east-1 as the default region for S3.
+        if service_name == 's3' and region_name is None:
+            region_name = 'us-east-1'
+        return self._proxy.construct_endpoint(service_name, region_name)
 
+
+class EndpointResolverChain(BaseEndpointResolver):
+    """Resolves endpoints using a chain of endpoint resolvers"""
+    def __init__(self, endpoint_resolvers):
+        self.endpoint_resolvers = endpoint_resolvers
+
+    def list_endpoint_names(self, service_name, partition_name='aws',
+                            allow_non_regional=False):
+        for resolver in self.endpoint_resolvers:
+            endpoints = resolver.list_endpoint_names(
+                service_name, partition_name, allow_non_regional)
+            if endpoints:
+                return endpoints
+        return []
+
+    def construct_endpoint(self, service_name, region_name):
+        for resolver in self.endpoint_resolvers:
+            endpoint = resolver.construct_endpoint(service_name, region_name)
+            if endpoint:
+                return endpoint
+        return None
+
+
+class LazyPartitionEndpointResolver(object):
```

Comment:
Yeah making the partitions into one file is going to simplify a lot of this logic